homeassistant/components/tplink/config_flow.py

```diff
@@ -52,15 +78,88 @@ async def _async_handle_discovery(self, host: str, mac: str) -> FlowResult:
         for progress in self._async_in_progress():
             if progress.get("context", {}).get(CONF_HOST) == host:
                 return self.async_abort(reason="already_in_progress")
-
+        credentials = await get_stored_credentials(self.hass)
         try:
-            self._discovered_device = await self._async_try_connect(
-                host, raise_on_progress=True
+            await self._async_try_discover_connect(
+                host, credentials, raise_on_progress=True
             )
+        except AuthenticationException:
+            self._last_failed_discovery_credentials = credentials
+            return await self.async_step_discovery_auth_confirm()
         except SmartDeviceException:
             return self.async_abort(reason="cannot_connect")
+
         return await self.async_step_discovery_confirm()
 
+    async def async_step_discovery_auth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Dialog that informs the user that auth is required."""
+        assert self._discovered_device is not None
+        errors = {}
+        host = self.context[CONF_HOST]
+        credentials = await get_stored_credentials(self.hass)
+
+        if credentials != self._last_failed_discovery_credentials:
+            httpx_asyncclient = create_async_httpx_client(self.hass, verify_ssl=False)
+            try:
```

Comment:
I think we have already a device instance at this point, even when it failed to auth? If so, we should not construct a new instance but rather update the credentials and re-try.

Reply:
The discovered device does not have the ha `httpx_asyncclient` which is why I had to recreate it.  You can't pass `httpx_client` to the `discover()` function because each device needs it's own client and you can't pass a generator function because the ha `httpx_asyncclient` has to be created on the event loop and discovery creates devices on the `datagram_received` thread.

Reply:
Why are we using threads with asyncio?

Reply:
> Why are we using threads with asyncio?

We're not, I was having a brain fade because `datagram_received` is not an async function but of course it's on the event loop.  The discover function in the python-kasa library will now generate a HassHttpXAsyncClient for each discovered device.

> I think we have already a device instance at this point, even when it failed to auth? If so, we should not construct a new instance but rather update the credentials and re-try.

The `device` api doesn't currently support simply updating credentials and retrying.  It's doable but for now this piece of code will take the `ConnectionParams` from the constructed device, update the credentials, and pass them to the `SmartDevice.connect`.

Reply:
We need to avoid creating httpx clients on the fly since it will create an SSLContext which is expensive and does I/O in the event loop (and leaks memory on some versions of openssl)

For onvif we do https://github.com/hunterjm/python-onvif-zeep-async/blob/040cdef7b37ee1c345bccc001d3ba091e39c98a1/onvif/client.py#L44 to avoid creating multiple ssl contexts

But its better to pass in the httpx session from HASS to the lib

Reply:
As far as I can tell the `verify_ssl=False` parameter is doing the same thing as onvif.  Maybe someone copied it from onvif to be core HASS util function? https://github.com/home-assistant/core/blob/dev/homeassistant/util/ssl.py#L64.

We can't use the shared HASS client because the cookies need to be unique (and cleared) but we're using the `create_async_httpx_client` from `homeassistant.helpers.httpx_client` library so should be ok no?

Reply:
That should be fine, just need to make sure discovery is also creating it that way as well

Reply:
I've been switching a lot, so I can't remember if we actually create an httpx client or not during discovery so will have to dig it on that one a bit.

I did notice we could use a future instead of an asyncio.event in the discovery code, which would be a little more efficient since there's only a single waiter but it's probably not running enough that it matters

Reply:
If we are using verify=False it will create a new ssl context every time.  If we use the helper it doesn't have that problem.  Verify= also accepts an ssl context

Reply:
I see a St client being created in the library, but I haven't finished tracing back if that code ever called from Home Assistant as I keep getting sidetracked

Reply:
Ok so to check my understanding:

1. You think the hass httpx helper creates a new default ssl context each time even with `verify_ssl=False` so better to create it once in the integration and pass it in. 

2. You think the integration might be creating its own AsyncClient in discovery somehow. I don’t think it is but I can double check.

Reply:
Thanks for summarizing.


> Ok so to check my understanding:
> 
> 1. You think the hass httpx helper creates a new default ssl context each time even with `verify_ssl=False` so better to create it once in the integration and pass it in.

The helper should be safe as I fixed that in #90191.  Its only a problem if we are creating AsyncClient directly and passing `verify=False`

> 2. You think the integration might be creating its own AsyncClient in discovery somehow. I don’t think it is but I can double check.
That's the concern

If both of these lines are never hit from Home Assistant that all is good and we don't have an issue. If we call either of them, we need to pass the client instead.

https://github.com/python-kasa/python-kasa/blob/510aea7207111ef25750ed854ecd4d916801f72c/kasa/klaptransport.py#L140
https://github.com/python-kasa/python-kasa/blob/510aea7207111ef25750ed854ecd4d916801f72c/kasa/aestransport.py#L106